### PR TITLE
feat: add `schema_of_json` function

### PR DESCRIPTION
**Pull request by kentkr**
initial implementation of [`schema_of_json`](https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.functions.schema_of_json.html)

covers the basic implementation with two known gaps:
 - options handling is incomplete
 - array type inference only validates the first element

note: spark uses a custom json parser (loosely json5-like with exceptions), full compatibility may require extending `serde_json` or `serde_json5`

### CODE REVIEW

File: crates/sail-function/src/scalar/json/schema_of_json.rs [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+        remove_leading_zeros(string.to_string())
+    } else {
+        string.to_string()
+    }
+}
+
+#[expect(clippy::unwrap_used)]
+fn remove_leading_zeros(string: String) -> String {
+    let re = Regex::new(r"\b0+([1-9]\d*)").unwrap();
+    re.replace_all(string.as_str(), "$1").to_string()
```
[the remark this answers is not in the document]
Comment (kentkr) [reply crates/sail-function/src/scalar/json/schema_of_json.rs]: The implementation is fine, we're not outputting the json, just inferring types. I added a BDD test to check.

File: crates/sail-function/src/scalar/json/schema_of_json.rs [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+    };
+    let rows = downcast_arg!(&args[0], StringArray);
+    let options = if let Some(arg) = args.get(1) {
+        let map_array = downcast_arg!(arg, MapArray);
+        SparkSchemaOfJsonOptions::default().map_to_options(map_array)?
+    } else {
+        SparkSchemaOfJsonOptions::default()
+    };
+    let type_ddl = infer_json_schema_type(rows.value(0), &options)?;
+    Ok(Arc::new(StringArray::from(vec![type_ddl])))
```
[the remark this answers is not in the document]
Comment (kentkr) [reply crates/sail-function/src/scalar/json/schema_of_json.rs:157]: The literal will never be null, the function call would be invalid.

File: python/pysail/tests/spark/function/features/schema_of_json.feature [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        from rows
+        """
+      Then query error Expected a literal value
+
+    Scenario: schema of json null input
+      When query
+        """
+        SELECT schema_of_json(null) AS result
+        """
+      Then query error Execution error:.*found invalid arg types: \[Null\]
```
[the remark this answers is not in the document]
Comment (linhr) [reply python/pysail/tests/spark/function/features/schema_of_json.feature]: Yeah it's better to have loose error message matching here. This doesn't seem to cause CI failure though.

File: crates/sail-plan/src/function/scalar/json.rs [hunk 4]
```diff
@@ -59,7 +59,7 @@ pub(super) fn list_built_in_json_functions() -> Vec<(&'static str, ScalarFunctio
         ("json_array_length", F::unary(json_array_length)),
         ("json_object_keys", F::unary(json_object_keys)),
         ("json_tuple", F::unknown("json_tuple")),
-        ("schema_of_json", F::unknown("schema_of_json")),
+        ("schema_of_json", F::udf(SparkSchemaOfJson::new())),
```
Comment (linhr) [inline crates/sail-plan/src/function/scalar/json.rs:62]: Let's register this function in `crates/sail-execution/src/codec.rs`.

File: crates/sail-function/src/scalar/json/schema_of_json.rs [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+    }
+
+    fn get_keys_values_from_map(inner_struct: StructArray) -> Result<(StringArray, StringArray)> {
+        let (keys, values) = match inner_struct.data_type() {
+            DataType::Struct(fields) => {
+                let key_type = fields[0].data_type();
+                let value_type = fields[1].data_type();
+                if key_type == &DataType::Utf8 && value_type == &DataType::Utf8 {
+                    let keys = downcast_array::<StringArray>(inner_struct.column(0));
+                    let values = downcast_array::<StringArray>(inner_struct.column(1));
```
[the remark this answers is not in the document]
Comment (linhr) [reply crates/sail-function/src/scalar/json/schema_of_json.rs:259]: This comment is probably incorrect since CI is successful.

File: crates/sail-function/src/scalar/json/schema_of_json.rs [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+        remove_leading_zeros(string.to_string())
+    } else {
+        string.to_string()
+    }
+}
+
+#[expect(clippy::unwrap_used)]
+fn remove_leading_zeros(string: String) -> String {
+    let re = Regex::new(r"\b0+([1-9]\d*)").unwrap();
+    re.replace_all(string.as_str(), "$1").to_string()
```
Comment (linhr) [inline crates/sail-function/src/scalar/json/schema_of_json.rs]: I'm not sure if this is correct. If we have a JSON string `{"a_01b": 01}`, would we correctly retain the field name `a_01b`?

In general, we can't use regex for such a task since we need a parser. But I can see that modifying the JSON parser is out of scope for this PR, so I'd prefer to just return an "unsupported" error if `allow_numeric_leading_zeros` is true.

File: crates/sail-function/src/scalar/json/schema_of_json.rs [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+        let map_array = downcast_arg!(arg, MapArray);
+        SparkSchemaOfJsonOptions::default().map_to_options(map_array)?
+    } else {
+        SparkSchemaOfJsonOptions::default()
+    };
+    let type_ddl = if rows.is_empty() {
+        return Err(DataFusionError::Execution(
+            "No value passed into input".to_string(),
+        ));
+    } else if rows.value(0).is_empty() {
```
[the remark this answers is not in the document]
Comment (linhr) [reply crates/sail-function/src/scalar/json/schema_of_json.rs:152]: This comment is probably incorrect.

File: crates/sail-function/src/scalar/json/schema_of_json.rs (hunk 6 above)
Comment (kentkr) [reply crates/sail-function/src/scalar/json/schema_of_json.rs]: Oh good catch. A leading 0 on the key would result in an incorrect output.

File: crates/sail-plan/src/function/scalar/json.rs (hunk 4 above)
Comment (kentkr) [reply crates/sail-plan/src/function/scalar/json.rs:62]: @linhr done, let me know if I missed anything

File: crates/sail-function/src/scalar/json/schema_of_json.rs [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+        Ok(self)
+    }
+
+    fn get_keys_values_from_map(inner_struct: StructArray) -> Result<(StringArray, StringArray)> {
+        let (keys, values) = match inner_struct.data_type() {
+            DataType::Struct(fields) => {
+                let key_type = fields[0].data_type();
+                let value_type = fields[1].data_type();
+                if key_type == &DataType::Utf8 && value_type == &DataType::Utf8 {
+                    let keys = downcast_array::<StringArray>(inner_struct.column(0));
```
[the remark this answers is not in the document]
Comment (kentkr) [reply crates/sail-function/src/scalar/json/schema_of_json.rs:258]: yeesh aight clanker you're right

File: crates/sail-execution/src/codec.rs [hunk 9]
```diff
@@ -1792,6 +1780,10 @@ impl PhysicalExtensionCodec for RemoteExecutionCodec {
                 let udf = SparkFromCSV::new(Arc::from(session_timezone));
                 return Ok(Arc::new(ScalarUDF::from(udf)));
             }
+            UdfKind::SparkSchemaOfJson(gen::SparkSchemaOfJsonUdf {}) => {
+                let udf = SparkSchemaOfJson::new();
+                return Ok(Arc::new(ScalarUDF::from(udf)));
+            }
```
Comment (linhr) [inline crates/sail-execution/src/codec.rs]: The encode/decode logic does not match.

You don't need to update `.proto`. Since this UDF does not have arguments, you need to add the UDF to the `match name { ... }` below.

File: crates/sail-function/src/scalar/json/schema_of_json.rs [hunk 10]
```diff
@@ -111,7 +109,23 @@ impl ScalarUDFImpl for SparkSchemaOfJson {
     fn coerce_types(&self, arg_types: &[DataType]) -> Result<Vec<DataType>> {
         Self::validate_args_len(arg_types)?;
         Self::validate_arg_types(arg_types)?;
-        Ok(arg_types.to_vec())
+        let mut coerce_to = vec![DataType::Utf8];
+        if arg_types.len() > 1 {
+            // Forcer le map<k,v> → map<Utf8, Utf8>
```
Comment (linhr) [inline crates/sail-function/src/scalar/json/schema_of_json.rs]: The comment needs to be in English.

Comment (kentkr) [reply crates/sail-function/src/scalar/json/schema_of_json.rs]: hahha oops

File: crates/sail-execution/src/codec.rs [hunk 11]
```diff
@@ -1879,6 +1879,9 @@ impl PhysicalExtensionCodec for RemoteExecutionCodec {
             "json_object_keys" | "json_keys" => {
                 Ok(sail_function::scalar::json::json_object_keys_udf())
             }
+            "spark_schema_of_json" | "schema_of_json" => {
+                Ok(Arc::new(ScalarUDF::from(SparkSchemaOfJson::new())))
+            }
```
Comment (linhr) [inline crates/sail-execution/src/codec.rs:1884]: The encode method is missing. It's kind of unfortunate to have multiple back-and-forth for the small changes in this file. Let's take a look at how other functions are handled in this file to really understand the setup.

**Comment by linhr** [comment]
Sorry I just realized this PR is completing the same tasks as #1225 that was created and reviewed in January.

I feel this one can go first since it handles the map options to some extend. But #1225 handles array type inference in a better way and has more extensive BDD tests. So maybe we can port the core UDF logic and tests from #1225 after this PR is merged. @davidlghellin What do you think?

Sorry for not realizing the duplicated efforts beforehand.

**Comment by davidlghellin** [comment]
> I feel this one can go first since it handles the map options to some extend. But #1225 handles array type inference in a better way and has more extensive BDD tests. So maybe we can port the core UDF logic and tests from #1225 after this PR is merged. @davidlghellin What do you think?

Sounds good to me! I think merging #1540 first makes sense since. After that, I'm happy to help port the array type inference logic and expand the BDD test coverage from #1225. I've been working on edge cases in this area so I can take care of that follow-up.

File: crates/sail-execution/src/codec.rs (hunk 11 above)
Comment (kentkr) [reply crates/sail-execution/src/codec.rs:1884]: @linhr Good call, and sorry for the back-and-forth. I should have taken a broader look at the file before submitting. I'm still getting my footing with Rust and proto, so I wasn't confident what the expected encode/decode pattern was. I'll steal from the #1225 implementation and get it all updated in one go

**Comment by kentkr** [comment]
No worries @linhr, this was easier than `from_json` so I learned more; it was worth it.

@davidlghellin that'd be great!

**Comment by kentkr** [comment]
@linhr can you trigger the skipped checks please?

**Comment by linhr** [comment]
Oh sorry I was meant to merge this PR but somehow it didn't go through.

The skipped tests are fine. We'll monitor the build on the `main` branch once the PR is merged.

Thanks again for your contribution!

**Comment by davidlghellin** [comment]
https://github.com/lakehq/sail/issues/219
